flow/src/test/java/com/sequenceiq/flow/core/chain/FlowChainHandlerTest.java

```diff
@@ -0,0 +1,152 @@
+
+    private static final String TRIGGER_VALUE = "triggerValue";
+
+    private static final String SELECTOR = "simpleSelector";
+
+    private static final String SIMPLE_VALUE = "chainKey";
+
+    @Mock
+    private FlowChains flowChains;
+
+    @Mock
+    private FlowLogService flowLogService;
+
+    @Mock
+    private FlowStatCache flowStatCache;
+
+    @Mock
+    private FlowChainOperationTypeConfig flowChainOperationTypeConfig;
+
+    @InjectMocks
+    private FlowChainHandler underTest;
+
+    @Captor
+    private ArgumentCaptor<FlowTriggerEventQueue> eventQueueCaptor;
+
+    @Test
+    void testRestoreFlowChainNotFound() {
+        when(flowLogService.findFirstByFlowChainIdOrderByCreatedDesc(FLOW_CHAIN_ID)).thenReturn(Optional.empty());
+        underTest.restoreFlowChain(FLOW_CHAIN_ID);
+        verifyNoInteractions(flowChains, flowStatCache, flowChainOperationTypeConfig);
+    }
+
+    @ParameterizedTest(name = "use Jackson = {0}")
+    @ValueSource(booleans = { false, true })
+    void testRestoreFlowChain(boolean useJackson) {
+        FlowChainLog flowChainLog = new FlowChainLog(FLOW_CHAIN_TYPE, FLOW_CHAIN_ID, PARENT_FLOW_CHAIN_ID,
+                "{\"@type\":\"java.util.concurrent.ConcurrentLinkedQueue\",\"@items\":[{\"@type\":\"" +
+                        Simple.class.getName() + "\",\"resourceId\":123,\"key\":\"" + SIMPLE_VALUE + "\"}]}",
+                useJackson
+                        ? "[{\"@type\":\"" +
+                            Simple.class.getName() + "\",\"resourceId\":123,\"key\":\"" + SIMPLE_VALUE + "\"}]"
+                        : null,
+                USER_CRN,
+                "{\"@type\":\"" + Simple.class.getName() + "\",\"key\":\"" + TRIGGER_VALUE + "\"}",
+                useJackson
+                        ? "{\"@type\":\"" + Simple.class.getName() + "\",\"key\":\"" + TRIGGER_VALUE + "\"}"
+                        : null);
+        when(flowLogService.findFirstByFlowChainIdOrderByCreatedDesc(FLOW_CHAIN_ID)).thenReturn(Optional.of(flowChainLog));
+        when(flowChainOperationTypeConfig.getFlowTypeOperationTypeMap()).thenReturn(Map.of(FLOW_CHAIN_TYPE, OperationType.PROVISION));
+        underTest.restoreFlowChain(FLOW_CHAIN_ID);
+
+        verify(flowChains).putFlowChain(eq(FLOW_CHAIN_ID), eq(PARENT_FLOW_CHAIN_ID), eventQueueCaptor.capture());
+        FlowTriggerEventQueue eventQueue = eventQueueCaptor.getValue();
+        assertThat(eventQueue.getFlowChainName()).isEqualTo(FLOW_CHAIN_TYPE);
+        assertThat(eventQueue.getParentFlowChainId()).isEqualTo(PARENT_FLOW_CHAIN_ID);
+        assertThat(eventQueue.getTriggerEvent()).isInstanceOf(Simple.class);
+        assertThat(((Simple) eventQueue.getTriggerEvent()).getKey()).isEqualTo(TRIGGER_VALUE);
+        assertThat(eventQueue.getQueue()).contains(new Simple(SELECTOR, SIMPLE_VALUE));
+        verify(flowStatCache).putByFlowChainId(FLOW_CHAIN_ID, 123L, "PROVISION", true);
+    }
+
+    static class Simple implements Selectable {
```

Comment:
Nit: Perhaps rename to `SimpleSelectable`.

Reply:
Well... :D
This is used in the test class only, but whatever.

Reply:
Thanks :smile: